src/main/java/dev/morling/onebrc/CalculateAverage_armandino.java

```diff
@@ -18,173 +18,111 @@
+        private final Map<Integer, Stats> map = new HashMap<>(2048);
+        private final SampleHolder sample = new SampleHolder();
 
-            for (int i = 0; i < numChunks; i++) {
-                long start = 0;
-                long end = chunkSize;
+        private List<Stats> process(final ByteBuffer bb) {
+            // not sure if this is faster... from royvanrijn
+            bb.order(ByteOrder.nativeOrder());
 
-                if (i > 0) {
-                    start = chunks[i - 1].end + 1;
-                    end = Math.min(start + chunkSize, fileSize);
+            while (bb.hasRemaining()) {
+                byte b;
+                while ((b = bb.get()) != SEMICOLON) {
+                    sample.pushKey(b);
                 }
-
-                end = end == fileSize ? end : seekNextNewline(channel, end);
-                chunks[i] = new Chunk(start, end);
-            }
-            return chunks;
-        }
-
-        private static long seekNextNewline(final FileChannel channel, final long end) throws IOException {
-            var bb = ByteBuffer.allocate(MAX_KEY_LENGTH);
-            channel.position(end).read(bb);
-
-            for (int i = 0; i < bb.limit(); i++) {
-                if (bb.get(i) == NL) {
-                    return end + i;
+                while ((b = bb.get()) != NL) {
+                    if (b == MINUS) {
+                        sample.negative = true;
+                    }
+                    else if (b != DOT) {
+                        sample.pushMeasurement(b);
+                    }
                 }
+                addSample(sample);
+                sample.reset();
             }
-
-            throw new IllegalStateException("Couldn't find next newline");
+            return list;
         }
 
-        private void process(final ByteBuffer bb) {
-            final var sample = new Sample();
-            var isKey = true;
-
-            for (long i = 0, sz = bb.limit(); i < sz; i++) {
-
-                final byte b = bb.get();
+        private void addSample(final SampleHolder sample) {
+            Stats stats = map.get(sample.keyHash);
 
-                if (b == SEMICOLON) {
-                    isKey = false;
-                }
-                else if (b == NL) {
-                    isKey = true;
-                    addSample(sample);
-                    sample.reset();
-                }
-                else if (isKey) {
-                    sample.pushKey(b);
-                }
-                else if (b == DOT) {
-                    // skip
-                }
-                else if (b == MINUS) {
-                    sample.sign = -1;
-                }
-                else {
-                    sample.pushMeasurement(b);
-                }
+            if (stats == null) {
+                stats = new Stats(new String(sample.keyBytes, 0, sample.keyLength, UTF_8));
+                map.put(sample.keyHash, stats);
```

Comment:
This seems prone to hash collisions? There could be multiple names with the same hash and results would be wrong. I suppose that actually was an issue with the initial implementation already.

Reply:
@gunnarmorling thanks for the feedback. This should be equivalent to using `String` as a key if ingot this right. Maybe we need a test for collisions?